Add Zamba (#30950)
huggingface/transformers: docs/source/en/model_doc/zamba.md

# Zamba

Zamba is a large language model (LLM) trained by Zyphra, and made available under an Apache 2.0 license. Please see the [Zyphra Hugging Face](https://huggingface.co/Zyphra) repository for model weights.

## Model details

Zamba-7B-v1 is a hybrid between state-space models (Specifically [Mamba](https://github.com/state-spaces/mamba)) and transformer. Zamba consists of a Mamba backbone with a shared transformer block every 6 Mamba layers. We came to this architecture after a series of ablations at small scales, where we found it proved highly effective at combining the inference efficiency of Mamba with the minimal amount of attention required to maintain performance and expressivity. Zambaand was trained using next-token prediction and uses the [Mistral v0.1 tokenizer](https://huggingface.co/mistralai/Mistral-7B-v0.1).  Zamba-7B-v1 was pre-trained on 1T tokens of text and code data. We then performed an annealing phase over 50B high quality tokens.

<img src="https://github.com/Zyphra/HF-zamba/blob/main/zamba-arch.png" width=40% height=40% />

## Quick start

> ### Presequities

Reviewer: ```suggestion
### Prerequities
```

Zamba requires you use `transformers` version 4.39.0 or higher:
```bash
pip install transformers>=4.39.0
```

In order to run optimized Mamba implementations on a CUDA device, you first need to install `mamba-ssm` and `causal-conv1d`:
```bash
pip install mamba-ssm causal-conv1d>=1.2.0
```

You can run the model not using the optimized Mamba kernels, but it is **not** recommended as it will result in significantly higher latency. 

> To run on CPU, please specify `use_mamba_kernels=False` when loading the model using ``AutoModelForCausalLM.from_pretrained``.

Reviewer: Is this true? I don't think we do support that argument in `from_pretrained` currently

Author: Yes, we tested this. `use_mamba_kernels` is a parameter in `ZambaConfig` that defaults to `True`. We tested that passing `use_mamba_kernels=False` in `from_pretrained` sets the forward pass to the one using the pure Python implementation.

Reviewer: Indeed, it should be dealt here: https://github.com/huggingface/transformers/blob/6b11f89c6b3d15655a2264df73e453cf9692c62e/src/transformers/modeling_utils.py#L3229C19-L3229C25 all good !

> To run on CPU, please specify `use_mamba_kernels=False` when loading the model using ``AutoModelForCausalLM.from_pretrained``.

Reviewer: ```suggestion
To run on CPU, please specify `use_mamba_kernels=False` when loading the model using `AutoModelForCausalLM.from_pretrained`.
```

# Zamba

Zamba is a large language model (LLM) trained by Zyphra, and made available under an Apache 2.0 license. Please see the [Zyphra Hugging Face](https://huggingface.co/collections/zyphra/) repository for model weights.

## Model details

Zamba-7B-v1 is a hybrid between state-space models (Specifically [Mamba](https://github.com/state-spaces/mamba)) and transformer, and was trained using next-token prediction. Zamba uses a shared transformer layer after every 6 mamba blocks. It uses the [Mistral v0.1 tokenizer](https://huggingface.co/mistralai/Mistral-7B-v0.1). We came to this architecture after a series of ablations at small scales. Zamba-7B-v1 was pre-trained on 1T tokens of text and code data.

<img src=https://github.com/user-attachments/assets/c2cff209-b901-483c-87aa-774b82a0769f width=30% height=40% />

## Quick start

### Presequities

> Jamba requires you use `transformers` version 4.39.0 or higher:

Reviewer 2: ```suggestion
Zamba requires you use `transformers` version 4.45.0 or higher:
```

```bash

> pip install transformers>=4.39.0

Reviewer 2: ```suggestion
pip install transformers>=4.45.0
```

specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->
# Zamba

Zamba is a large language model (LLM) trained by Zyphra, and made available under an Apache 2.0 license. Please see the [Zyphra Hugging Face](https://huggingface.co/collections/zyphra/) repository for model weights.

## Model details

Zamba-7B-v1 is a hybrid between state-space models (Specifically [Mamba](https://github.com/state-spaces/mamba)) and transformer, and was trained using next-token prediction. Zamba uses a shared transformer layer after every 6 mamba blocks. It uses the [Mistral v0.1 tokenizer](https://huggingface.co/mistralai/Mistral-7B-v0.1). We came to this architecture after a series of ablations at small scales. Zamba-7B-v1 was pre-trained on 1T tokens of text and code data.

<img src=https://github.com/user-attachments/assets/c2cff209-b901-483c-87aa-774b82a0769f width=30% height=40% />

## Quick start

### Presequities

> Zamba requires you use `transformers` version 4.45.0 or higher:

Reviewer 2: ```suggestion
Zamba requires you use `transformers` version 4.46.0 or higher:
```
but we can do another release soon!

```bash
pip install transformers>=4.45.0
```

In order to run optimized Mamba implementations, you first need to install `mamba-ssm` and `causal-conv1d`:
```bash
pip install mamba-ssm causal-conv1d>=1.2.0
```
You also have to have the model on a CUDA device.

You can run the model not using the optimized Mamba kernels, but it is **not** recommended as it will result in significantly lower latencies. In order to do that, you'll need to specify `use_mamba_kernels=False` when loading the model.

## Inference

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("Zyphra/Zamba-7B-v1")
model = AutoModelForCausalLM.from_pretrained("Zyphra/Zamba-7B-v1", device_map="auto", torch_dtype=torch.bfloat16)

input_text = "A funny prompt would be "
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")

outputs = model.generate(**input_ids, max_new_tokens=100)
print(tokenizer.decode(outputs[0]))
```

> 

Reviewer 2: We usually add `This model was contributed by xxx` and the username (hf handle) of the person that opened the PR

Reviewer 3: resolved in https://github.com/Zyphra/transformers_zamba/commit/a46a26ba48410cc1fae91fe06c2c9a138b3f8252
